docs: add READMEs for Slurm integration interface implementations (#239)
charmed-hpc/slurm-charms: internal/sackd-interface/README.md

# `charmed-slurm-sackd-interface`

## Usage

This package provides the integration interface implementation for the `sackd` interface. It enables `sackd`
(Slurm authentication and credential kiosk daemon) applications to receive controller data from `slurmctld`,

> including authentication secrets and controller addresses.

Reviewer: ```suggestion
This package provides the integration interface implementation for the `sackd` interface. It enables `sackd` (Slurm authentication and credential kiosk daemon) applications to exchange data with `slurmctld`, including authentication secrets and controller addresses.
```

To install, add `charmed-slurm-sackd-interface` to your Python dependencies _pyproject.toml_.
Then in your Python code, import as:

```python
from charmed_slurm_sackd_interface import (
    SackdProvider,
    SackdRequirer,
    controller_ready,
)
```

## Direction

The `sackd` interface implements a provider/requirer pattern.
The Provider is the `sackd` application that receives controller data from `slurmctld`.

> The Requirer is the `slurmctld` application that provides authentication secrets and controller addresses to `sackd`.

Reviewer: ```suggestion
The `sackd` interface implements the provider/requirer integration pattern. The provider-side application, `sackd`, provides the integration endpoint that the requirer-side application, `slurmctld`, needs. `slurmctld` then supplies authentication secrets and controller addresses to `sackd` through the application databag associated with the integration instance between `sackd` and `slurmctld`.
```

```mermaid
flowchart TD

>     Requirer -- auth_secret_id, controllers --> Provider

Reviewer: ```suggestion
flowchart TD
    B("sackd\n(provider)") -- endpoint --> A("slurmctld\n(requirer)")
    A("slurmctld\n(requirer)") -- auth_secret_id, controllers --> B("sackd\n(provider)")
```

Author: `endpoint` shouldn't be added into the integration databag content since it's not a "key" that the Slurm charms insert into the databag. An endpoint is defined in the _metadata.yaml_/_charmcraft.yaml_ file and not inside the integration databag: https://documentation.ubuntu.com/juju/3.6/reference/application/#application-endpoint.

The arrows should only contain the physical information that is exchanged between the applications through the integration databags. `auth_secret_id` and `controllers` for example.

Reviewer: Ah ok, it wasn't clear that the diagram was only showing databag-related items. Could we add a quick statement or indicator in the diagram itself making that clearer? Without the endpoint in the diagram, there's no 'initiation' starting from the provider; I'm trying to avoid diagrams that only have requirer -> provider.

Author: Hmm... do we actually need an "initiation" explicitly added to the diagram? It's rather par-for-the-course for existing interface documentation to include unidirectional diagrams if it's only the provider or requirer putting data into the integration databag. See existing interface documentation for example: https://github.com/canonical/charmlibs/tree/main/interfaces/fiveg_n3/interface/v0.

The "Direction" diagram should only present the data that is transferred over the integration as that's the most important information charm developers typically want; imho everything else is superfluous and should be left for documentation outside the README. The "Behavior" section can allude to the expected flow, but shouldn't meticulously explain how Juju integration events are triggered. I don't think we should try to re-explain how Juju integrations work here.

Reviewer: Happy to be overruled on these, especially if there's a "standard" diagram. I'll just reiterate that it would help if it was more clear that the diagram is showing explicitly what's in the databag in that case. Agreed we don't need to try to re-explain how integrations work - any that's common understanding feel free to ignore me.

Author: I'm just trying to avoid having us be apophenic in the sense that we attempt to attribute higher meaning to visualizations in diagrams that mean nothing to people who don't have the entire context of this conversation. It's the same quagmire we got stuck in with the reference diagrams where we overstructured the diagram and its meaning got completely lost on folks external to our core group because they didn't have the context we did.

This visualization is more or less intended to communicate "hey the provider doesn't send anything to the requirer, but the requirer sends this information"; it's not meant to communicate deeper-level things like the expected integration event lifecycle or the user actions that trigger these events. Those types of concepts are better served by higher-fidelity sequence or state diagrams. Most of that information is already captured in specifications like [UHPC001](https://github.com/charmed-hpc/specs/blob/main/specs/UHPC%20001%20-%20Dynamic%20nodes%20in%20Charmed%20HPC/uhpc-001.md).

```

## Behavior

The `slurmctld` requirer provides controller data (authentication secret and controller addresses) to
the `sackd` provider through the relation databag. The `sackd` provider validates that required fields

> are present before considering the integration ready.

Reviewer: This section could be fully merged with the Direction section above; maybe just a slight tweak to above about validating the required fields?

## Usage

This package provides the integration interface implementation for the `sackd` interface.
It enables charmed applications providing the `sackd` service (Slurm auth and cred kiosk daemon)
to exchange Slurm management data with charmed applications that require the `sackd` service.

The `sackd` requirer provides controller data (authentication secret and controller addresses) to
the `sackd` provider through the integration databag. The `sackd` provider validates that required fields

> are present before considering the integration ready.

Reviewer: ```suggestion
The `sackd` requirer supplies controller data (authentication secret and controller addresses) to
the `sackd` provider through the integration databag. The `sackd` provider validates that required fields
are present before considering the integration ready.
```

> ## Integration data

Reviewer: Minor nit: this doesn't need to have a whole header for just a source link

Author: Sure, I can wordsmith the headers a bit since `Source` also provides information about the "Behavior" of the provider and requirer.

Author: I changed the header to "Example integration data" and removed the `Source` link as I think the example data is still useful for understanding how the data is represented in Juju. Presumably if our reader is going through the README, then they already have access to the source code if they want to learn more about how the interface is actually implemented 😅
